File: web_cabinet/ai/prompts/weekly_brief.py
"""Промпт для еженедельного брифинга фермы (MVP-N17)."""
from __future__ import annotations

from typing import Any

WEEKLY_BRIEF_SYSTEM = """\
Ты — старший ИИ-аналитик GenomeAI. Каждую неделю ты готовишь развёрнутый \
executive-брифинг для руководителя молочной фермы — стратегический отчёт \
уровня инвестора.

ЯЗЫК: Строго русский. Профессиональный аналитический стиль. \
Конкретные цифры и факты. Никаких общих слов без данных. \
Никакого marketing bullshit.

ОБЯЗАТЕЛЬНАЯ СТРУКТУРА ОТВЕТА (строго валидный JSON):
{
  "title": "Недельный отчёт: ДД-ДД месяц ГГГГ",
  "executive_summary": "3-5 предложений: ключевые изменения недели с цифрами. Обязательно упомяни главный успех и главную проблему.",
  "sections": [
    {
      "heading": "Продуктивность",
      "narrative": "2-4 параграфа с анализом надоев, трендов, динамики по группам. Сравнение с прошлой неделей.",
      "highlights": ["Ключевой момент 1 с цифрой", "Ключевой момент 2"],
      "evidence_ids": ["evt_id_из_контекста"]
    }
  ],
  "key_recommendations": [
    {
      "recommendation": "Конкретное действие — что, кому, когда",
      "priority": "high",
      "rationale": "Почему важно — конкретный факт из данных [evidence_id]",
      "expected_outcome": "Что улучшится и насколько",
      "affected_entities": ["cow_id или group_id"]
    }
  ],
  "anomalies_detected": [
    {
      "description": "Аномалия с показателем и величиной отклонения",
      "severity": "critical",
      "evidence_id": "evt_id_из_контекста"
    }
  ],
  "kpi_table": {
    "avg_milk_yield_kg": {"value": 28.4, "prev_period": 29.1, "delta_pct": -2.4, "unit": "кг/день"},
    "scc_thousands": {"value": 312, "prev_period": 287, "delta_pct": 8.7, "unit": "тыс/мл"}
  }
}

ОБЯЗАТЕЛЬНЫЕ СЕКЦИИ: Продуктивность, Воспроизводство, Здоровье (минимум 3).
Добавляй Кормление только если есть данные.

РЕКОМЕНДАЦИИ: строго 3–7. Каждая с rationale из данных и expected_outcome.

АНОМАЛИИ: только реальные отклонения из farm_context. \
Severity: critical = немедленное действие, warning = в течение суток, info = к сведению.

EVIDENCE GROUNDING: ОБЯЗАТЕЛЬНО. evidence_ids в sections и evidence_id в anomalies — \
ТОЛЬКО существующие ID из farm_context. Вымышленные ID — ЗАПРЕЩЕНЫ.

ОБЪЁМ: суммарно narrative 800–1500 слов.

ЗАПРЕЩЕНО: вымышленные event_id, рекомендации без обоснования из данных, \
пустые секции, повторение одного факта в разных секциях.\
"""


def build_weekly_brief_message(context: Any, week_start: str, week_end: str) -> str:
    """Строит user message для еженедельного брифинга."""
    context_text = context.to_text() if hasattr(context, "to_text") else str(context)
    return (
        f"Период анализа: {week_start} — {week_end}\n\n"
        f"<farm_context>\n{context_text}\n</farm_context>\n\n"
        f"Подготовь полноценный недельный брифинг за период {week_start} — {week_end}.\n"
        f"Верни строго валидный JSON согласно схеме из системного промпта.\n"
        f"Используй только event_id из farm_context выше."
    )

File: web_app/lib/api/weekly-brief.ts
import { apiFetch } from '@/lib/api/client';

export interface WeeklyBriefPeriod {
  start: string;
  end: string;
}

export interface BriefSection {
  heading: string;
  narrative: string;
  highlights: string[];
  evidence_ids: string[];
}

export interface KeyRecommendation {
  recommendation: string;
  priority: 'high' | 'medium' | 'low';
  rationale: string;
  expected_outcome: string;
  affected_entities: string[];
}

export interface WeeklyAnomaly {
  description: string;
  severity: 'critical' | 'warning' | 'info';
  evidence_id: string;
}

export interface KpiEntry {
  value: number;
  prev_period: number | null;
  delta_pct: number | null;
  unit: string;
}

export interface WeeklyBrief {
  brief_id: string;
  farm_id: string;
  period: WeeklyBriefPeriod;
  generated_at_utc: string;
  title: string;
  executive_summary: string;
  sections: BriefSection[];
  key_recommendations: KeyRecommendation[];
  anomalies_detected: WeeklyAnomaly[];
  kpi_table: Record<string, KpiEntry>;
  generation_model: string;
  generation_tokens: { input: number; output: number };
}

export async function fetchWeeklyBrief(farmId = 'INV_FARM_001'): Promise<WeeklyBrief> {
  return apiFetch<WeeklyBrief>(
    `/api/ai/weekly-brief/latest?farm_id=${encodeURIComponent(farmId)}`,
  );
}

export async function generateWeeklyBrief(
  farmId = 'INV_FARM_001',
  startDate?: string,
  endDate?: string,
): Promise<WeeklyBrief> {
  return apiFetch<WeeklyBrief>('/api/ai/weekly-brief', {
    method: 'POST',
    body: JSON.stringify({
      farm_id: farmId,
      start_date: startDate ?? '',
      end_date: endDate ?? '',
      force_regenerate: true,
    }),
  });
}

export function weeklyBriefPdfUrl(briefId: string, farmId = 'INV_FARM_001'): string {
  return `/api/backend/api/ai/weekly-brief/${briefId}/pdf?farm_id=${encodeURIComponent(farmId)}`;
}
